contracts/strategies/morpho/MetaMorphoStrategy.sol

```diff
@@ -149,11 +203,208 @@ contract MetaMorphoStrategy is ERC4626Strategy {
+      ).supplyRewardTokensPerYear / _oneReward;
+  }
+
+  /// @notice calculate how much of vault `_add` amount will be added to this market
+  /// @param _mmVault metamorpho vault
+  /// @param _targetMarketId target market id
+  /// @param _supplyQueueLen supply queue length
+  /// @param _add amount of liquidity to add
+  function _calcMarketAdd(
+    IMMVault _mmVault,
+    bytes32 _targetMarketId,
+    uint256 _supplyQueueLen,
+    uint256 _add
+  ) internal view returns (uint256) {
+    uint256 _assetsSuppliedByVault;
+    uint184 _marketCap;
+    bytes32 _currMarketId;
+    IMorpho.Market memory _market;
+    IMorpho.Position memory _pos;
+
+    // loop throuh supplyQueue, starting from the first market, and see how much will
+    // be deposited in target market
+    for (uint256 i = 0; i < _supplyQueueLen; i++) {
+      _currMarketId = _mmVault.supplyQueue(i);
+      _market = MORPHO_BLUE.market(_currMarketId);
+      _pos = MORPHO_BLUE.position(_currMarketId, address(_mmVault));
+      _assetsSuppliedByVault = _pos.supplyShares * _market.totalSupplyAssets / _market.totalSupplyShares;
+      // get max depositable amount for this market
+      (_marketCap,,) = _mmVault.config(_currMarketId);
+      uint256 _maxDeposit;
+      if (_assetsSuppliedByVault < uint256(_marketCap)) {
+        _maxDeposit = uint256(_marketCap) - _assetsSuppliedByVault;
+      }
+      // If this is the target market, return the current _add value, eventually
+      // reduced to the max depositable amount
+      if (_currMarketId == _targetMarketId) {
+        if (_add > _maxDeposit) {
+          _add = _maxDeposit;
+        }
+        break;
+      }
+      // If this is not the target market, check if we can deposit all the _add amount
+      // in this market, otherwise continue the loop and subtract the max depositable
+      if (_add > _maxDeposit) {
+        _add -= _maxDeposit;
+      } else {
+        _add = 0;
+        break;
+      }
+    }
+
+    return _add;
+  }
+
+  /// @notice calculate how much of vault `_sub` amount will be removed from target market
+  /// @param _mmVault metamorpho vault
+  /// @param _targetMarketId target market id
+  /// @param _withdrawQueueLen withdraw queue length
+  /// @param _sub liquidity to remove
+  function _calcMarketSub(
+    IMMVault _mmVault, 
+    bytes32 _targetMarketId,
+    uint256 _withdrawQueueLen,
+    uint256 _sub
+  ) internal view returns (uint256) {
+    IMorpho.Market memory _market;
+    bytes32 _currMarketId;
+    // loop throuh withdrawQueue, and see how much will be redeemed in target market
+    for (uint256 i = 0; i < _withdrawQueueLen; i++) {
+      _currMarketId = _mmVault.withdrawQueue(i);
+      _market = MORPHO_BLUE.market(_currMarketId);
+      // get available liquidity for this market
+      uint256 _availableLiquidity;
+      if (_market.totalBorrowAssets < _market.totalSupplyAssets) {
+        _availableLiquidity = _market.totalSupplyAssets - _market.totalBorrowAssets;
+      }
```

Comment:
```suggestion
      uint256 _availableLiquidity = _market.totalSupplyAssets - _market.totalBorrowAssets;
```

This is one of the strongest invariants of Morpho Blue (`totalSupplyAssets > totalBorrowAssets`)!

Reply:
I was doing this as I'm currently not using `zeroFloorSub` but if it's not needed than I will remove it

Reply:
changed https://github.com/Idle-Labs/idle-tranches/pull/85/commits/f4d7f21d4942dc2cf273c9a1218315ef246e7a0d